lib/benchee/benchmark/measure/memory.ex

```diff
@@ -44,17 +55,22 @@ defmodule Benchee.Benchmark.Measure.Memory do
   defp return_memory({memory_usage, result}) when memory_usage < 0, do: {nil, result}
   defp return_memory({memory_usage, result}), do: {memory_usage, result}
 
-  defp measure_memory(fun, tracer) do
-    word_size = :erlang.system_info(:wordsize)
-    {:garbage_collection_info, heap_before} = Process.info(self(), :garbage_collection_info)
-    result = fun.()
-    {:garbage_collection_info, heap_after} = Process.info(self(), :garbage_collection_info)
-    mem_collected = get_collected_memory(tracer)
+  defp measure_memory(fun, tracer, parent) do
+    :erlang.garbage_collect()
```

Comment:
Shouldn't this GC be made after you receive the :ready_to_begin? How else will you know the memory when starting the benchmark?

Also, by doing the final GC after the send you are counting the {:result, result} tuple creation as part of the memory allocated by the benchmark.

Reply:
Good point I think :thinking: Thanks for checking in, @devonestes what do you think?

Reply:
hmm, yes I see what you mean and that should work.

I just thought of something else, when doing the final GC, I think it will reclaim the `fun` term. This will also make your measurements a little bit off.

Reply:
Interesting! I can include a reference to that term after the final GC just to be sure. Maybe I'll have `measure_memory/2` return `fun` just so it doesn't get collected.

Reply:
Now you just have to include the mbuf_size, bin_vheap_size and bin_old_vheap_size in the memory measurements :) To keep track of any terms allocated as fragments and off heap binaries created.

Speaking of the old heap, just so that you are aware. With the current measurements a term that is allocated and lives long enough to be on the old heap is counted 3 times in the memory measurements, once when initially allocated, once when moved on the young heap and then again when moved to the old heap. Not sure what it is that you want to measure, just thought you should know.

Reply:
Oh boy - more stuff to measure! Yeah, at first we wanted to cover the most basic application that would get us mostly there. But you're totally right, it would be more precise and correct to also include fragments and off-heap binaries in the total memory usage of a function.

I don't think we'll have issues with double counting because of movement from the young heap to old heap, since when we look at the data for a given GC event we're counting the total heap used by adding `heap_size` and `old_heap_size`. So, if something is only moved from the heap to the old heap, it would still have the same total. Only when something new is allocated to one of those two heaps or when something is garbage collected from them will that total change.

Basically our algorithm to find the total memory is: (heap_size + old_heap_size before GC) - (heap_size + old_heap_size after GC). If something is moved from the young heap to the old heap, that shouldn't affect our calculations.

What do you mean, though, when you say "initially allocated" and then moved to the young heap? When a term is allocated, is it not allocated directly on the young heap? I may be missing something important here!

Reply:
> What do you mean, though, when you say "initially allocated" and then moved to the young heap? When a term is allocated, is it not allocated directly on the young heap? I may be missing something important here!

I mean that a term is copied once on the young heap before it is copied onto the old heap. As described here, see especially the notes on high-watermark: https://github.com/erlang/otp/blob/master/erts/emulator/internal_doc/GarbageCollection.md#generational-garbage-collection

Reply:
Oh, yeah, I don't think that's going to be a problem.

That's a great document that I haven't seen before. Very helpful! Thanks for sharing the link!